### ISSUE TITLE
Invalid RA diagnostic error: expected 2 arguments, found 1

### ISSUE THREAD

**Issue by mckenfra** [rust-lang/rust-analyzer#17497]
`rust-analyzer version: 0.3.2011-standalone (2fd803cc1 2024-06-22)`
`rustc 1.78.0 (9b00956e5 2024-04-29)`
`VSCode, Version 0.3.2011, Server Version 0.3.2011-standalone (2fd803cc1 2024-06-22)`

While working on a rust project in VSCode, I noticed that RA was reporting errors with perfectly valid code. After stripping back all surrounding code, I have come up with the below minimal example that reproduces the error.

The error message I see in VSCode is: `expected 2 arguments, found 1` rust-analyzer [E0107](https://doc.rust-lang.org/stable/error_codes/E0107.html)

This occurs on the line `vec.push("hello");`

```rust
#![allow(dead_code)]
use core::ops::Deref;

trait Trait1 {
    type Assoc: Deref<Target = String>;
}
trait Trait2: Trait1 {
}
trait Trait3 {
    type T1: Trait1;
    type T2: Trait2;
    fn push(&self, x: bool, y: bool);
}

fn problem_method<T: Trait3>() {
    let mut vec = Vec::new();
    vec.push("hello"); // Rustc ok, RA errors (mismatched args)
}
```

**Comment by lnicola** [comment]
Fixed in #17511.
